### ISSUE TITLE
checkhealth: reports (non)-error with shada file location

### ISSUE THREAD

**Issue by lyu** [neovim/neovim#11046]
- `nvim --version`: 0.4.2 (from Archlinux official repo)
- `vim -u DEFAULTS` (version: ) behaves differently? Not applicable
- Operating system/version: Archlinux x86_64
- Terminal name/version: tilix 1.9.1
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

```
(make sure you have removed the shada file from its default location)
nvim -u NORC -c 'set shada+=n/tmp/nvim.shada' -c 'checkhealth'
```

### Actual behaviour
Although nvim does create nvim.shada under /tmp, it says
```
ERROR: shada file is not readable:
    /home/my_user_name/.local/share/nvim/shada/main.shada
```
in the health checks.
### Expected behaviour
Health checks pass, like v0.3.8 did.

**Comment by ivanalejandro0** [comment]
I'm experiencing the same issue on OSX.

**Comment by blueyed** [comment]
Via https://github.com/neovim/neovim/commit/5ae57b7e571ff8faefcbee84af24bf58ec097fa3 / https://github.com/neovim/neovim/pull/10327.

@justinmk 
Should probably check the parent dir when the file does not exist (`filewritable(fnamemodify(…, ':h')) == 2`)?

**Comment by blueyed** [comment]
```diff
diff --git i/runtime/autoload/health/nvim.vim w/runtime/autoload/health/nvim.vim
index c25f5ee64..247ecb242 100644
--- i/runtime/autoload/health/nvim.vim
+++ w/runtime/autoload/health/nvim.vim
@@ -42,9 +42,13 @@ function! s:check_config() abort
   endif
 
   let shadafile = (empty(&shadafile) || &shadafile ==# 'NONE') ? stdpath('data').'/shada/ma>
-  if !empty(shadafile) && (!filereadable(shadafile) || !filewritable(shadafile))
-    let ok = v:false
-    call health#report_error('shada file is not '.(filereadable(shadafile) ? 'writeable' : >
+  if !empty(shadafile)
+    let readable = filereadable(shadafile)
+    let writable = filewritable(shadafile)
+    if (!readable || !writable) && !filewritable(fnamemodify(shadafile, ':h')) == 2
+      let ok = v:false
+      call health#report_error('shada file is not '.(readable ? 'writeable' : 'readable').">
+    endif
   endif
 
   if ok
```

NOTE: `filereadable` might be false for an existing file, where then also a writable dir does not help.
Should use `glob()` then probably.

**Comment by blueyed** [comment]
As for the issue title: only `:checkhealth` complains, right?  The shada location should work, doesn't it?

**Comment by blueyed** [comment]
As for `checkhealth`: it probably also needs to parse the "shadafile" out of `&shada` if it is empty..

**Comment by blueyed** [comment]
As for the test case: it needs to be `--cmd 'set shada+=n/tmp/nvim5.shada'` (not `-c`), or `-i /tmp/nvim5.shada` - not sure if that's a regression then though.

**Comment by blueyed** [comment]
> vim -u DEFAULTS (version: ) behaves differently? Not applicable

Could use `viminfo` there, but not really necessary.

> Although nvim does create nvim.shada under /tmp, it says

Have not seen this before - so apparently is an issue with checkhealth then only - likely the parsing of `&shada` for an empty `&shadafile` should be done there.
But given the hardcoded default location there already, it might be better to use some API function for it altogether, to prevent issues like this (checks not matching actual behavior).

**Comment by blueyed** [comment]
But please confirm that there is no regression you're seeing outside of `:checkhealth`, please.

**Comment by lyu** [comment]
@blueyed 
I executed: `nvim -u NORC --cmd 'set shada+=n/tmp/nvim.shada' test.txt`, searched for something, then exit.

If I execute the above command again and press `n` when I am in nvim, it still remembers the search string.

If I remove `/tmp/nvim.shada`, then execute the above command, nvim forgets the search string.

Does this confirm that there is no regression other than `:checkhealth` showing the wrong thing?

**Comment by blueyed** [comment]
@lyu yes - with a new shada file no history is available, and it picks up the existing one.

@ivanalejandro0 
Have you found the issue yourself, or only reproduced it from here?

Do you (both) use a custom shada location by default?  (I am interested in use cases)

**Comment by ivanalejandro0** [comment]
I found the problem myself and came to the issues on the repo to see if anyone else was experiencing the same issue.

I found it by running the `:checkhealth` right after a `brew upgrade neovim`.

I'm experiencing the same as @lyu .
The command `vim -u NORC --cmd 'set shada+=n/tmp/nvim.shada' -c 'checkhealth' test.txt` reports:

```
  - ERROR: shada file is not readable:
    /Users/ivan/.local/share/nvim/shada/main.shada
```

But closing the editor and opening it again preserves the search string.

I'm using a custom shada file location, on my `~/.config/nvim/init.vim` I have
```vim
set shada+='1000,n~/.config/nvim/dirs/shada
```

I have the same problem (on `:checkhealth`) if I `set shadafile`.

**Comment by lyu** [comment]
Yes, I also use a custom shada file location by default.

**Comment by justinmk** [comment]
just need to adjust checkhealth. +1 for https://github.com/neovim/neovim/issues/11046#issuecomment-532290267

**Comment by lyu** [comment]
Sorry for the late reply, I can confirm that the patch provided by @blueyed works for me.

**Comment by lyu** [comment]
Problem fixed after the merge of #13291
